File: week2/task1.py
mrt_stations = {
    "Songshan": {"Nanjing Sanmin": 1},
    "Nanjing Sanmin": {"Songshan": 1, "Taipei Arena": 1},
    "Taipei Arena": {"Nanjing Sanmin": 1, "Nanjing Fuxing": 1},
    "Nanjing Fuxing": {"Taipei Arena": 1, "Songjiang Nanjing": 1},
    "Songjiang Nanjing": {"Nanjing Fuxing": 1, "Zhongshan": 1},
    "Zhongshan": {"Songjiang Nanjing": 1, "Ximen": 1},
    "Ximen": {"Zhongshan": 1, "Xiaonanmen": 1},
    "Xiaonanmen": {"Ximen": 1, "Chiang Kai-shek Memorial Hall": 1},
    "Chiang Kai-shek Memorial Hall": {"Xiaonanmen": 1, "Guting": 1},
    "Guting": {"Chiang Kai-shek Memorial Hall": 1, "Taipower Building": 1},
    "Taipower Building": {"Guting": 1, "Gongguan": 1},
    "Gongguan": {"Taipower Building": 1, "Wanlong": 1},
    "Wanlong": {"Gongguan": 1, "Jingmei": 1},
    "Jingmei": {"Wanlong": 1, "Dapinglin": 1},
    "Dapinglin": {"Jingmei": 1, "Qizhang": 1},
    "Qizhang": {"Dapinglin": 1, "Xindian City Hall": 1, "Xiaobitan": 1},
    "Xiaobitan": {"Qizhang": 1},
    "Xindian City Hall": {"Qizhang": 1, "Xindian": 1},
    "Xindian": {"Xindian City Hall": 1}
}

def find_and_print(messages, current_station):
    closest_friend = None
    min_distance = float('inf')

    def station_distance(start, end):
        if start == end:
            return 0
        visited = set()
        queue = [(start, 0)]

        while queue:
            station, distance = queue.pop(0)

            if station == end:
                return distance

            visited.add(station)

            for neighbor, dist in mrt_stations[station].items():
                if neighbor not in visited:
                    queue.append((neighbor, distance + dist))
        
        return float('inf')

    def station_name(message):
        for station in mrt_stations.keys():
            if station in message:
                return station
        return None

    for friend, message in messages.items():
        station = station_name(message)
        if station:
            distance = station_distance(current_station, station)
            if distance < min_distance:
                min_distance = distance
                closest_friend = friend

    if closest_friend:
        print(closest_friend)


messages = {
    "Bob": "I'm at Ximen MRT station.",
    "Mary": "I have a drink near Jingmei MRT station.",
    "Copper": "I just saw a concert at Taipei Arena.",
    "Leslie": "I'm at home near Xiaobitan station.",
    "Vivian": "I'm at Xindian station waiting for you."
}

find_and_print(messages, "Wanlong")  # print Mary
find_and_print(messages, "Songshan")  # print Copper
find_and_print(messages, "Qizhang")  # print Leslie
find_and_print(messages, "Ximen")  # print Bob
find_and_print(messages, "Xindian City Hall")  # print Vivian

File: week2/task1.js
const mrtStations = [
    "Songshan",
    "Nanjing Sanmin",
    "Taipei Arena",
    "Nanjing Fuxing",
    "Songjiang Nanjing",
    "Zhongshan",
    "Ximen",
    "Xiaonanmen",
    "Chiang Kai-shek Memorial Hall",
    "Guting",
    "Taipower Building",
    "Gongguan",
    "Wanlong",
    "Jingmei",
    "Dapinglin",
    "Qizhang",
    "Xindian City Hall",
    "Xindian"
];

function findAndPrint(messages, currentStation) {
    let closestFriend = null;
    let minDistance = Infinity;

    //找到訊息中的站名
    function stationName(message) {
        for (let station of mrtStations) {
            if (message.includes(station)) {
                return station;
            }
        }
        return null;
    }
    //計算兩個站之間的距離
    function stationDistance(station1, station2) {
        let index1 = mrtStations.indexOf(station1);
        let index2 = mrtStations.indexOf(station2);
        return Math.abs(index1 - index2);
    }
    //判斷哪位朋友最近
    for (let [friend, message] of Object.entries(messages)) {
        let station = stationName(message);
        if (station) {
            let distance = stationDistance(station, currentStation);
            if (distance < minDistance) {
                minDistance = distance;
                closestFriend = friend;
            }
        }
    }
    if (closestFriend) {
        console.log(closestFriend);
    } 
}

const messages = {
    "Bob": "I'm at Ximen MRT station.",
    "Mary": "I have a drink near Jingmei MRT station.",
    "Copper": "I just saw a concert at Taipei Arena.",
    "Leslie": "I'm at home near Xiaobitan station.",
    "Vivian": "I'm at Xindian station waiting for you."
};

findAndPrint(messages, "Wanlong"); // print Mary
findAndPrint(messages, "Songshan"); // print Copper
findAndPrint(messages, "Qizhang"); // print Leslie
findAndPrint(messages, "Ximen"); // print Bob
findAndPrint(messages, "Xindian City Hall"); // print Vivian
